### PR TITLE
Add pre-commit

### PR DESCRIPTION
**Pull request by cbrnr**
This should include all style and other code quality checks. Currently, I've added Black, Ruff, and Codespell. I'll start with their default configs, but will add custom settings (probably in `pyproject.toml`).

To do:

- [x] Add custom config
- [x] Remove "codespell and flake" workflow as required
- [ ] Make Pre-commit workflow required before running other workflows
- [ ] Wait for #10641 before enabling Black
- [ ] Add other hooks that are not available in Codespell, Ruff, and Black

### CODE REVIEW

**Comment by cbrnr** [comment]
It seems like the `tomli` dependency lines can be removed with the next release of Codespell (https://github.com/codespell-project/codespell/issues/2644).

**Comment by larsoner** [comment]
New codespell released https://pypi.org/project/codespell/2.2.3/

**Comment by cbrnr** [comment]
That was quick, thanks for the heads up! I will remove the tomli lines.

**Comment by larsoner** [comment]
... tomli is still needed for <3.11, though, assuming you want toml support

**Comment by cbrnr** [comment]
Yes, but I'm running pre-commit on Python 3.11.

**Comment by cbrnr** [comment]
Now I wonder why Codespell found these errors in our Azure job but neither in the old GitHub Action nor the new Pre-Commit one:

```
mne/decoding/tests/test_csp.py:263: shuold ==> should
mne/forward/forward.py:1264: matirx ==> matrix
mne/utils/check.py:850: instatiating ==> instantiating
doc/changes/0.24.inc:296: identitiy ==> identity
```

OK, the last one is clear, because I didn't add `.inc` files, but the other three?

**Comment by drammock** [comment]
> Now I wonder why Codespell found these errors in the [Azure job](https://dev.azure.com/mne-tools/mne-python/_build/results?buildId=24096&view=logs&j=c1ec3add-0fcd-52ea-530a-328457bbfb2e&t=370d4985-0b6a-5123-801e-618bab9d7575&l=24) but neither in the old GitHub Action nor the new Pre-Commit one:

my suspicion would be that the azure job is somehow getting different dictionary settings or using a newer codespell version.

**Comment by drammock** [comment]
crossref to #11547

**Comment by cbrnr** [comment]
OK, I fixed it here as well.

**Comment by cbrnr** [comment]
But this begs the question: why do we run Codespell twice?

**Comment by drammock** [comment]
> But this begs the question: why do we run Codespell twice?

no idea.

**Comment by cbrnr** [comment]
I'll remove it then.

**Comment by cbrnr** [comment]
OK, it seems like we are checking against a very small subset of [Pydocstyle rules](https://beta.ruff.rs/docs/rules/#pydocstyle-d). Would it be a good idea to try to enable as many rules as possible, meaning that we should try to fix our docstrings instead of ignoring most rules? Some of them should be pretty simple (missing docstring in public module, etc.), but others might not be fixable (e.g. D417 "Missing argument descriptions in the docstring" everywhere we use template values like `%(verbose)s`). In any case, this will go in a separate PR after this one.

**Comment by cbrnr** [comment]
OK, I don't have more time to work on this in the next days, so if someone wants to take over please feel free! I was trying to specify separate exclude directories for `ruff.pydocstyle`, but it seems like this can only be set globally. No idea how to fix this ATM.

**Comment by larsoner** [comment]
>> For which, like doc/ versus mne/?

> I would need different Ruff settings for different directories.

I think that's okay if it's `doc/` and `mne/` (is it?) then we can put the doc ones in `doc/` for example and the mne ones in the root or so

**Comment by cbrnr** [comment]
Or use different command line args in the pre-commit YAML, we'll see what's easier/less confusing.

**Comment by larsoner** [comment]
Which directories do we need to check? Just `mne/` and `doc/`?

**Comment by cbrnr** [comment]
For `flake8 `we currently have:

    exclude = __init__.py,constants.py,fixes.py,resources.py,*doc/auto_*,*doc/_build*,build/*

And for `pydocstyle`:

    match_dir = ^(?!\.|doc|tutorials|examples|logo|icons).*$
    match = (?!tests/__init__\.py|fixes).*\.py

**Comment by larsoner** [comment]
@cbrnr I'd like to have this sooner rather than later because the time it takes to `make flake` locally is starting to bug me :)

Want me to push commits to finish or can you do it in the next couple of days?

**Comment by cbrnr** [comment]
Thanks @larsoner, please feel free to finish this up. I still don't know how to run different configurations for ruff-specific tools (i.e. pydocstyle should only be run in `mne/`, but everything else in all directories). I also don't know how you would like to handle pydocstyle-related warnings – I'd prefer to fix them, but most warnings seem to be outside of `mne/` anyway.

I'm happy to review the final PR of course!

**Comment by larsoner** [comment]
Okay @cbrnr I think this is working, feel free to take a look. I think we can wait on the other points in the top comment. For the GH Actions job dependencies we should first in a separate PR put all our pytest actions in one YAML (which we only had separate for back when you couldn't restart just failed runs) along with this new precommit one. Then we can make the pytest ones depend on the style job very easily.

**Comment by larsoner** [comment]
@drammock could you also look and make sure you're okay with these changes? Basically:

1. Replace `flake8` and `pydocstyle` with `ruff` which does the same work much faster
2. Add `.pre-commit-config` where people can do `pre-commit install --install-hooks` and have `git commit` reject when style is not satisfied. Or they don't have to do this and the CIs will yell at them, as usual.
3. Alias `make flake` and `make pep` to use the `pre-commit` config
4. Update contrib docs to mention `pre-commit`

**Comment by larsoner** [comment]
I think we can ignore the SphinxWindows failures... they don't seem related.

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+      name: ruff mne
+      files: ^mne/
+
+# Ruff tutorials and examples
+- repo: https://github.com/charliermarsh/ruff-pre-commit
+  rev: v0.0.262
+  hooks:
+    - id: ruff
+      name: ruff tutorials and examples
+      args: ["--ignore=D103,D400"]
```
Comment (drammock) [inline .pre-commit-config.yaml:25]:
```suggestion
      # D103: missing docstring in public function
      # D400: docstring first line must end with period
      args: ["--ignore=D103,D400"]
```

File: doc/install/contributing.rst [hunk 2]
```diff
@@ -694,11 +694,16 @@ Adhere to standard Python style guidelines
 All contributions to MNE-Python are checked against style guidelines described
 in `PEP 8`_. We also check for common coding errors (such as variables that are
 defined but never used). We allow very few exceptions to these guidelines, and
-use tools such as pep8_, pyflakes_, and flake8_ to check code style
+use tools such as ruff_ to check code style
 automatically. From the :file:`mne-python` root directory, you can check for
-style violations by running::
+style violations by first installing pre-commit hook
```
Comment (drammock) [inline doc/install/contributing.rst:699]:
```suggestion
style violations by first installing our pre-commit hook::
```

File: pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,35 @@
+[tool.codespell]
+ignore-words = "ignore_words.txt"
+uri-ignore-words-list = "bu"
+builtin = "clear,rare,informal,names,usage"
+skip = "doc/references.bib"
+
+[tool.ruff]
+select = ["E", "F", "W", "D"]
+exclude = ["__init__.py", "constants.py", "fixes.py", "resources.py"]
+ignore = ["D100", "D104", "D107", "D413"]
```
Comment (drammock) [inline pyproject.toml:10]:
```suggestion
ignore = [
    "D100", # Missing docstring in public module
    "D104", # Missing docstring in public package
    "D107", # Missing docstring in __init__
    "D413", # Missing blank line after last section
]
```

Comment (drammock) [reply pyproject.toml:10]: if we're excluding `__init__.py` why do we need D107?  Is the `exclude` only excluding the main `__init__.py` for the whole package, and the `ignore` still applys to submodule `__init__.py`s?

File: pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    "mne.utils.copy_function_doc_to_method_doc",
+    "mne.utils.copy_doc",
+    "mne.utils.deprecated"
+]
+
+[tool.ruff.per-file-ignores]
+"tutorials/time-freq/10_spectrum_class.py" = ["E501"]
+"mne/datasets/*/*.py" = ["D103"]
+"mne/utils/tests/test_docs.py" = ["D101", "D411", "D410", "D414"]
+"examples/*/*.py" = ["D205"]
```
Comment (drammock) [inline pyproject.toml:26]:
```suggestion
"examples/*/*.py" = [
    "D205", # 1 blank line required between summary line and description
]
```

File: pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    "property",
+    "setter",
+    "mne.utils.copy_function_doc_to_method_doc",
+    "mne.utils.copy_doc",
+    "mne.utils.deprecated"
+]
+
+[tool.ruff.per-file-ignores]
+"tutorials/time-freq/10_spectrum_class.py" = ["E501"]
+"mne/datasets/*/*.py" = ["D103"]
```
Comment (drammock) [inline pyproject.toml:24]:
```suggestion
"mne/datasets/*/*.py" = [
    "D103", # Missing docstring in public function
]
```

File: pyproject.toml [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+ignore-decorators = [
+    "property",
+    "setter",
+    "mne.utils.copy_function_doc_to_method_doc",
+    "mne.utils.copy_doc",
+    "mne.utils.deprecated"
+]
+
+[tool.ruff.per-file-ignores]
+"tutorials/time-freq/10_spectrum_class.py" = ["E501"]
```
Comment (drammock) [inline pyproject.toml:23]:
```suggestion
"tutorials/time-freq/10_spectrum_class.py" = [
    "E501" # line too long
]
```

File: pyproject.toml [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+"mne/datasets/*/*.py" = ["D103"]
+"mne/utils/tests/test_docs.py" = ["D101", "D411", "D410", "D414"]
```
Comment (drammock) [inline pyproject.toml:25]:
```suggestion
"mne/utils/tests/test_docs.py" = [
    "D101", # Missing docstring in public class
    "D410", # Missing blank line after section
    "D411", # Missing blank line before section
    "D414", # Section has no content
]
```

**Review by drammock — COMMENTED** [review]
GitHub is highlighting my suggested changes in red, but AFAIK [it is valid TOML](https://toml.io/en/v1.0.0#array) to include line comments in this way. Is `pyproject.toml` more narrowly restricted?

File: pyproject.toml (hunk 3 above)
Comment (larsoner) [reply pyproject.toml:10]: I'll test locally after committing your suggestions

**Comment by larsoner** [comment]
Locally your changes seemed to work fine so I've committed all of them, and also removed D107. It was probably made redundant at some point and we never noticed. Will merge once everything except SphinxWindows comes back green, thanks @cbrnr !

**Comment by cbrnr** [comment]
Thank you!

**Pull request merged**